libraries/common/array.effekt

```diff
@@ -185,6 +209,282 @@ def sum(list: Array[Int]): Int = {
+///
+/// O(N)
+def any[A](arr: Array[A]) { predicate: A => Bool }: Bool = {
+  not(arr.all { x => not(predicate(x)) })
+}
+
+/// Count elements satisfying the predicate.
+///
+/// O(N)
+def count[A](arr: Array[A]) { predicate: A => Bool }: Int = {
+  var count = 0
+  arr.foreach { x =>
+    if (predicate(x)) count = count + 1
+  }
+  count
+}
+
+/// Take first n elements.
+/// Returns a new array.
+///
+/// O(min(n, size))
+def take[A](arr: Array[A], n: Int): Array[A] = {
+  val len = min(max(0, n), arr.size)
+  val result = allocate[A](len)
+  each(0, len) { i => 
+    result.unsafeSet(i, arr.unsafeGet(i))
+  }
+  result
+}
+
+// Drop first n elements.
+/// Returns a new array.
+/// 
+/// O(size - n)
+def drop[A](arr: Array[A], n: Int): Array[A] = {
+  val start = min(max(0, n), arr.size)
+  val len = arr.size - start
+  val result = allocate[A](len)
+  each(0, len) { i =>
+    result.unsafeSet(i, arr.unsafeGet(i + start))
+  }
+  result
+}
+
+/// Return a slice from start (inclusive) to end (exclusive).
+/// Returns a new array.
+///
+/// O(N)
+def slice[A](arr: Array[A], start: Int, end: Int): Array[A] = {
+  val validStart = min(max(0, start), arr.size)
+  val validEnd = min(max(validStart, end), arr.size)
+  val len = validEnd - validStart
+  val result = allocate[A](len)
+  each(0, len) { i =>
+    result.unsafeSet(i, arr.unsafeGet(i + validStart))
+  }
+  result
+}
+
+/// Zip two arrays into an array of pairs.
+/// Length is minimum of input lengths.
+///
+/// O(min(N, M))
+def zip[A, B](this: Array[A], other: Array[B]): Array[(A, B)] = {
+  val len = min(this.size, other.size)
+  val result = allocate[(A, B)](len)
+  each(0, len) { i =>
+    result.unsafeSet(i, (this.unsafeGet(i), other.unsafeGet(i)))
+  }
+  result
+}
+
+/// Combine two arrays using given function.
+/// Length is minimum of input lengths.
+///
+/// O(min(N, M))
+def zipWith[A, B, C](this: Array[A], other: Array[B]) { f: (A, B) => C }: Array[C] = {
+  val len = min(this.size, other.size)
+  val result = allocate[C](len)
+  each(0, len) { i =>
+    result.unsafeSet(i, f(this.unsafeGet(i), other.unsafeGet(i)))
+  }
+  result
+}
+
+/// Get the _first_ index of the last occurance of `elem`, if there is any.
+/// 
+/// O(N)
+def indexOf[A](arr: Array[A], toFind: A) { eq: (A, A) => Bool }: Option[Int] = {
+  var found = None[Int]()
+  arr.foreachIndex { (idx, elem) =>
+    if (eq(elem, toFind) && found.isEmpty) {
+      found = Some(idx)
+    }
+  }
+  found
+}
+
+/// Get the _last_ index of the last occurance of `elem`, if there is any.
+/// 
+/// O(N)
+def lastIndexOf[A](arr: Array[A], toFind: A) { eq: (A, A) => Bool }: Option[Int] = {
+  var lastFound = None[Int]()
+  arr.foreachIndex { (idx, elem) =>
+    if (eq(elem, toFind)) {
+      lastFound = Some(idx)
+    }
+  }
+  lastFound
+}
+
+/// Perform binary search to find the index of a given element.
+/// 
+/// O(log N)
+def find[A](arr: Array[A], toFind: A) { ord: (A, A) => Ordering }: Option[Int] = {
```

Comment:
There are a few uses where `lowerBound` and `upperBound` are also helpful (for algorithmic stuff or for stuff like "if I have a list of offsets of newlines in a file and an offset, get me the offset of the start of the line)-- it's again binary search. But I don't think it's _too_ important.